### PR TITLE
chore: init ptl support with a basic mnist example [DET-4767]

### PR DESCRIPTION
**Pull request by hamidzr**
## Description

here is a draft of the basic building blocks for ptl adapter for our PytorchTrial

## Commentary (optional)

The goal of the pr is to review the basics and land a foundation to build on the rest of the PTL support on top of.

## Checklist

- [x] User-facing API changes need the "User-facing API Change" label.
- [x] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

### CODE REVIEW

File: harness/determined/experimental/pytorch_lightning/data_module.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
+import pytorch_lightning as ptl
+from determined.pytorch import DataLoader as DetDataLoader
+from torch.utils.data import DataLoader
+
+class DETLightningDataModule(ptl.LightningDataModule):
```
Comment (hamidzr) [inline harness/determined/experimental/pytorch_lightning/data_module.py:5]: todo: remove this sorta-support for now.

File: harness/determined/experimental/pytorch_lightning/adapter.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+def bail(msg: str = '', fail: bool = True):
+    msg = f'NotSupported: {msg}'
+    if fail:
+        raise TypeError(msg)
+    else:
+        # TODO use a logger
+        print(msg, file=sys.stderr)
+
+
+class DETLightningModule(ptl.LightningModule):
```
Comment (shiyuann) [inline harness/determined/experimental/pytorch_lightning/adapter.py:24]: Why do we need a separate class for it? We should take an instantiated lightning module from users rather than defining a separate lightning module as we discussed last time

File: harness/determined/experimental/pytorch_lightning/adapter.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+
+def check_compat(lm: DETLightningModule):
+    if len(signature(lm.training_step).parameters) > 2:
+        bail('`optimizer_idx` and `hiddens` are not supported.')
+    if len(signature(lm.validation_step).parameters) > 2:
+        bail('`dataloader_idx` is not supported.')
+
+
+class PTLAdapter(PyTorchTrial):
```
Comment (shiyuann) [inline harness/determined/experimental/pytorch_lightning/adapter.py:42]: Same. We don't seem to need a separate class for `PTLAdapter`. Instead, we should add a function `PytorchContext.wrap_lightning_module` as in the PRD

File: examples/experimental/mnist_ptl/ptl-image/Makefile [hunk 4]
```diff
@@ -0,0 +1,2 @@
+build:
```
[the remark this answers is not in the document]
Comment (hamidzr) [reply examples/experimental/mnist_ptl/ptl-image/Makefile:1]: sure thing. this is more a of helper for development and it should go away once we add ptl to our base image.

File: examples/experimental/mnist_ptl/.gitignore [hunk 5]
```diff
@@ -0,0 +1,2 @@
+MNIST/
```
[the remark this answers is not in the document]
Comment (hamidzr) [reply examples/experimental/mnist_ptl/.gitignore:1]: do you mean the whole .gitignore file or just `MNIST` directory?

File: harness/determined/experimental/pytorch_lightning/adapter.py (hunk 3 above)
Comment (hamidzr) [reply harness/determined/experimental/pytorch_lightning/adapter.py:42]: we updated the PRD.

File: examples/experimental/mnist_ptl/docker/Makefile [hunk 6]
```diff
@@ -0,0 +1,2 @@
+build:
+	docker build -f Dockerfile -t ptl-image .
```
Comment (shiyuann) [inline examples/experimental/mnist_ptl/docker/Makefile:2]: nonblocking: we should add PTL to the environment image. Should do that in [a separate ticket](https://determinedai.atlassian.net/browse/DET-4986).

File: examples/experimental/mnist_ptl/model_def.py [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    def __init__(self, context: PyTorchTrialContext) -> None:
+        super().__init__(context,
+                         ptl.LightningMNISTClassifier(get_hparam=context.get_hparam))
+
+        self.dm = ptl.MNISTDataModule()
+
+    def build_training_data_loader(self) -> DetDataLoader:
+        self.dm.setup()
+        dl = self.dm.train_dataloader()
+        # if we can support something like this more comprehensivly that'll give us a way to
```
Comment (shiyuann) [inline examples/experimental/mnist_ptl/model_def.py:26]: nit: remove this comment since it won't be useful to users.

File: examples/experimental/mnist_ptl/ptl.py [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+from torch.nn import functional as F
+from torchvision import datasets, transforms
+from torchvision.datasets import MNIST
+from pathlib import Path
+from determined.experimental.pytorch_lightning import HyperparamsProvider
+from typing import Optional
+import os
+
+
+# CHANGE exnted DETLightningModule
```
Comment (shiyuann) [inline examples/experimental/mnist_ptl/ptl.py:13]: nit: remove this comment.

File: harness/determined/experimental/pytorch_lightning/__init__.py [hunk 9]
```diff
@@ -0,0 +1,4 @@
+from determined.experimental.pytorch_lightning.adapter import (
+    HyperparamsProvider,
```
Comment (shiyuann) [inline harness/determined/experimental/pytorch_lightning/__init__.py:2]: blocking: it doesn't seem very necessary to have this class. Let's just use a dictionary.

File: examples/experimental/mnist_ptl/ptl.py [hunk 10]
```diff
@@ -0,0 +1,106 @@
+import torch
```
Comment (shiyuann) [inline examples/experimental/mnist_ptl/ptl.py:1]: blocking: let's not put the PTL support in experimental. There is actually nothing to experiment about.

File: harness/determined/experimental/pytorch_lightning/adapter.py [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+    if fail:
+        raise TypeError(msg)
+    else:
+        # QUESTION: do we have/want logger support.
+        print(msg, file=sys.stderr)
+
+
+def check_compat(lm: ptl.LightningModule):
+    if len(signature(lm.training_step).parameters) > 2:
+        bail("`optimizer_idx` and `hiddens` are not supported.")
```
Comment (shiyuann) [inline harness/determined/experimental/pytorch_lightning/adapter.py:25]: blocking: we should actually support `optimizer_idx`.

File: harness/determined/experimental/pytorch_lightning/adapter.py [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+import pytorch_lightning as ptl
+import torch
+
+from determined.pytorch import PyTorchTrial, PyTorchTrialContext
+
+TorchData = Union[Dict[str, torch.Tensor], Sequence[torch.Tensor], torch.Tensor]
+HyperparamsProvider = Callable[[str], Any]
+
+
+def bail(msg: str = "", fail: bool = True):
```
Comment (shiyuann) [inline harness/determined/experimental/pytorch_lightning/adapter.py:14]: blocking: please create an except class for invalid trial in `harness/determined/error.py` and just use `raise` directly since it is what is done across the codebase.

File: harness/determined/experimental/pytorch_lightning/adapter.py [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+
+
+def check_compat(lm: ptl.LightningModule):
+    if len(signature(lm.training_step).parameters) > 2:
+        bail("`optimizer_idx` and `hiddens` are not supported.")
+    if len(signature(lm.validation_step).parameters) > 2:
+        bail("`dataloader_idx` is not supported.")
+
+
+class PTLAdapter(PyTorchTrial):
```
Comment (shiyuann) [inline harness/determined/experimental/pytorch_lightning/adapter.py:30]: Will this class be instantiated by the trial controller directly? If so, then you shouldn't include `lightning_module` in the `__init__`. If it won't, then how you translate this class into a `PytorchTrial`?

**Review by shiyuann — COMMENTED** [review]
There are a couple of major things that need to change:

1.  we should just directly make the changes in the pytorch directory rather than put it in experimental.
2. `PTLAdapter` should be a separate class rather than inherit from `PytorchTrial` since `PTLAdapter` is not a superset of `PytorchTrial`.
3. The translation from `PytorchTrial` and `PTLAdapter` should happen in the place where `PytorchTrialController` loads the trial class.
4. we should support the optimizer index.

File: harness/determined/experimental/pytorch_lightning/__init__.py (hunk 9 above)
Comment (hamidzr) [reply harness/determined/experimental/pytorch_lightning/__init__.py:2]: hmm, dictionary like `Dict[str, any]` ? but it wouldn't match what `context.get_hparams` would it?

File: harness/determined/experimental/pytorch_lightning/adapter.py (hunk 13 above)
Comment (hamidzr) [reply harness/determined/experimental/pytorch_lightning/adapter.py:30]: we discussed the usecase.

**Comment by hamidzr** [comment]
> There are a couple of major things that need to change:
> 
> 1. we should just directly make the changes in the pytorch directory rather than put it in experimental.
> 2. `PTLAdapter` should be a separate class rather than inherit from `PytorchTrial` since `PTLAdapter` is not a superset of `PytorchTrial`.
> 3. The translation from `PytorchTrial` and `PTLAdapter` should happen in the place where `PytorchTrialController` loads the trial class.
> 4. we should support the optimizer index.

Thanks for the review.
we discussed 2 & 3, we'd keep those as is. I'll address 1 and 4 and the rest of the comments and expand ptl support in another PR.

File: harness/determined/experimental/pytorch_lightning/__init__.py (hunk 9 above)
Comment (shiyuann) [reply harness/determined/experimental/pytorch_lightning/__init__.py:2]: Why don't just pass in what `get_hparams` returns?

**Comment by hamidzr** [comment]
continuing the work here https://github.com/determined-ai/determined/pull/1959
